src/IQFeed.CSharpApiClient.Extensions/Lookup/Historical/Resample/TickMessageExtensions.cs

```diff
@@ -54,60 +55,70 @@ private static IEnumerable<HistoricalBar> ToHistoricalBarsAscending(this IEnumer
                 if (tick.LastSize == 0)
                     continue;
 
+                // Check if we need to close off the current bar before we do anything else
+                if (currentBar != null && tick.Timestamp >= nextTimestamp)
+                {
+                    // We've reached the end of this bar. Send it out! Setting the currentBar to null
+                    //   will cause a new bar to be generated on the first valid tick, below.
+                    currentBar.VWAP = currentBar.VWAP / currentBar.PeriodVolume;
+                    yield return currentBar;
+
+                    currentBar = null;
+                }
+
+                // Check for date change up here otherwise 'O' ticks will be ignored
+                //   if they are the first ticks of the day, which will make total volume incorrect
+                if (currentBarDate != null && tick.Timestamp.Date != currentBarDate)
+                {
+                    // Date has changed. Reset the numbers for the next one
+                    // Don't start a new bar here, because if the next tick is an 'O' we want to count the volume
+                    //  but not the tick, and if it's the only tick in a bar, then there shouldn't be a bar! (Apparently)
+                    totalVolume = 0;
+                    totalTrade = 0;
+                    currentBarDate = tick.Timestamp.Date;
+                }
+
                 totalVolume += tick.LastSize;
                 totalTrade += 1;
 
                 // to effect the price the trade must be C or E
                 if (tick.BasisForLast == 'O')
                     continue;
 
-                if (tick.Timestamp < nextTimestamp)
+                if (currentBar == null)
                 {
-                    if (tick.Last < currentBar.Low)
-                        currentBar.Low = tick.Last;
-
-                    if (tick.Last > currentBar.High)
-                        currentBar.High = tick.Last;
-
-                    currentBar.Close = tick.Last;
+                    // if we get here, we have a valid trade tick, and we need a new bar
+                    var currentTimestamp = tick.Timestamp.Trim(intervalTicks);
+                    currentBar = new HistoricalBar(
+                        currentTimestamp,
+                        tick.Last,
+                        tick.Last,
+                        tick.Last,
+                        tick.Last,
+                        0,
+                        0,
```

Comment:
minor... why not setting up the totaltrade + totalvolume? to stay consistent with other algos below? I know isnt required but.... maybe a little more explicit no?

Reply:
It was mainly because in a previous incarnation it would have led to an incorrect figure, as those totals were being accumulated below that setup. Now, simply because they are being set 10 lines later, and we can't set PeriodTrade/Volume because they are being accumulated later, so it seemed more understandable to leave them all as blank, and have them all taken care of further down.
I'm not precious about it, so will make the change if you want, but personally I think it's more obvious that they're set later if they're zero to start.

Reply:
okay I got it!